tsdb/tsi1/report.go

```diff
@@ -0,0 +1,374 @@
+		byTagKey:            false,
+	}
+}
+
+// ReportTSISummary is returned by a report-tsi Run() command and is used to access cardinality information
+type Summary struct {
+	TotalCardinality             int64
+	OrgCardinality               map[influxdb.ID]int64
+	BucketByOrgCardinality       map[influxdb.ID]map[influxdb.ID]int64
+	BucketMeasurementCardinality map[influxdb.ID]map[string]int64
+}
+
+func newSummary() *Summary {
+	return &Summary{
+		OrgCardinality:               make(map[influxdb.ID]int64),
+		BucketByOrgCardinality:       make(map[influxdb.ID]map[influxdb.ID]int64),
+		BucketMeasurementCardinality: make(map[influxdb.ID]map[string]int64),
+	}
+}
+
+// Run runs the report-tsi tool which can be used to find the cardinality
+// any org or bucket. Run returns a *ReportTSISummary, which contains maps for finding
+// the cardinality of a bucket or org based on its influxdb.ID
+func (report *ReportCommand) Run(print bool) (*Summary, error) {
+	report.start = time.Now()
+	report.Stdout = os.Stdout
+
+	if report.SeriesDirPath == "" {
+		report.SeriesDirPath = filepath.Join(report.DataPath, "_series")
+	}
+
+	sFile := tsdb.NewSeriesFile(report.SeriesDirPath)
+
+	// TODO: do we actually want the seriesfile logging?
+	// sFile.WithLogger(report.Logger)
+
+	if err := sFile.Open(context.Background()); err != nil {
+		report.Logger.Error("failed to open series")
+		return nil, err
+	}
+	defer sFile.Close()
+	report.sfile = sFile
+
+	indexPath := filepath.Join(report.DataPath, "index")
+	report.indexFile = NewIndex(sFile, NewConfig(), WithPath(indexPath))
+	if err := report.indexFile.Open(context.Background()); err != nil {
+		return nil, err
+	}
+	defer report.indexFile.Close()
+
+	summary, err := report.calculateOrgBucketCardinality()
+	if err != nil {
+		return nil, err
+	}
+
+	if print {
+		report.printCardinalitySummary(summary)
+	}
+
+	return summary, nil
+}
+
+type cardinality struct {
+	name  []byte
+	short []uint32
+	set   *tsdb.SeriesIDSet
+}
+
+func (c *cardinality) add(x uint64) {
+	if c.set != nil {
+		c.set.AddNoLock(tsdb.NewSeriesID(x))
+		return
+	}
+
+	c.short = append(c.short, uint32(x)) // Series IDs never get beyond 2^32
+
+	// Cheaper to store in bitmap.
+	if len(c.short) > useBitmapN {
+		c.set = tsdb.NewSeriesIDSet()
+		for i := 0; i < len(c.short); i++ {
+			c.set.AddNoLock(tsdb.NewSeriesID(uint64(c.short[i])))
+		}
+		c.short = nil
+		return
+	}
+}
+
+func (c *cardinality) cardinality() int64 {
+	if c == nil || (c.short == nil && c.set == nil) {
+		return 0
+	}
+
+	if c.short != nil {
+		return int64(len(c.short))
+	}
+	return int64(c.set.Cardinality())
+}
+
+func (report *ReportCommand) calculateCardinalities() error {
+	idx := report.indexFile
+	itr, err := idx.MeasurementIterator()
+	if err != nil {
+		return err
+	} else if itr == nil {
+		return nil
+	}
+	defer itr.Close()
+
+OUTER:
+	for {
+		name, err := itr.Next()
+		if err != nil {
+			return err
+		} else if name == nil {
+			break OUTER
+		}
```

Comment:
I would remove the `OUTER` label and just do a `return nil` here since there's nothing happening after the `for` loop (except the return).

Reply:
Also, I would move everything after this `if` block to a function called `calculateMeasurementCardinalities(name []byte)`. The loop body is pretty long and you'll be able to take advantage of `defer` for the `sitr.Close()` below.